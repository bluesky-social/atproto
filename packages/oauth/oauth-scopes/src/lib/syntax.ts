export type ParamValue = string | number | boolean

export type NeArray<T> = [T, ...T[]]

/**
 * Non-empty readonly array
 */
export type NeRoArray<T> = readonly [T, ...T[]]

export type ScopeStringFor<P extends string> =
  | P
  | `${P}:${string}`
  | `${P}?${string}`

/**
 * Allows to quickly check if a scope is for a specific resource.
 */
export function isScopeStringFor<P extends string>(
  value: string,
  prefix: P,
): value is ScopeStringFor<P> {
  if (value.length > prefix.length) {
    // First, check the next char is either : or ?
    const nextChar = value.charCodeAt(prefix.length)
    if (nextChar !== 0x3a /* : */ && nextChar !== 0x3f /* ? */) {
      return false
    }

    // Then check the full prefix
    return value.startsWith(prefix)
  } else {
    // value and prefix must be equal
    return value === prefix
  }
}

/**
 * Abstract interface that allows parsing various syntaxes into permission
 * representations.
 */
export interface ScopeSyntax<P extends string> {
  readonly prefix: P
  readonly positional?: ParamValue
  keys(): Iterable<string, void, unknown>
  getSingle(key: string): ParamValue | null | undefined
  getMulti(key: string): ParamValue[] | null | undefined
}
