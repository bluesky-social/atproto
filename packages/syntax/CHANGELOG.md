# @atproto/syntax

## 0.1.5

### Patch Changes

- [#1908](https://github.com/bluesky-social/atproto/pull/1908) [`3c0ef382`](https://github.com/bluesky-social/atproto/commit/3c0ef382c12a413cc971ae47ffb341236c545f60) Thanks [@gaearon](https://github.com/gaearon)! - prevent unnecessary throw/catch on uri syntax

## 0.1.4

### Patch Changes

- [#1788](https://github.com/bluesky-social/atproto/pull/1788) [`84e2d4d2`](https://github.com/bluesky-social/atproto/commit/84e2d4d2b6694f344d80c18672c78b650189d423) Thanks [@bnewbold](https://github.com/bnewbold)! - update license to "MIT or Apache2"

- Updated dependencies [[`84e2d4d2`](https://github.com/bluesky-social/atproto/commit/84e2d4d2b6694f344d80c18672c78b650189d423)]:
  - @atproto/common-web@0.2.3

## 0.1.3

### Patch Changes

- Updated dependencies [[`35d108ce`](https://github.com/bluesky-social/atproto/commit/35d108ce94866ce1b3d147cd0620a0ba1c4ebcd7)]:
  - @atproto/common-web@0.2.2

## 0.1.2

### Patch Changes

- Updated dependencies [[`41ee177f`](https://github.com/bluesky-social/atproto/commit/41ee177f5a440490280d17acd8a89bcddaffb23b)]:
  - @atproto/common-web@0.2.1

## 0.1.1

### Patch Changes

- [#1611](https://github.com/bluesky-social/atproto/pull/1611) [`b1dc3555`](https://github.com/bluesky-social/atproto/commit/b1dc355504f9f2e047093dc56682b8034518cf80) Thanks [@estrattonbailey](https://github.com/estrattonbailey)! - Fix imports in `README.md`.
