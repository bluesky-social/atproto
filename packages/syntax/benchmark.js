/* eslint-env node, commonjs */

const { validateNsid, validateNsidRegex } = require('.')

// $ node benchmark.js
// valid NSIDs {
//   parsed: 181.56524884700775,
//   regexp: 77.61082607507706,
//   optimized: 60.183539509773254
// }
// invalid NSIDs {
//   parsed: 128.7685609459877,
//   regexp: 108.75775015354156,
//   optimized: 53.196488440036774
// }

bench('valid NSIDs', true, [
  'com.example.foo',
  'o'.repeat(63) + '.foo.bar',
  'com.' + 'o'.repeat(63) + '.foo',
  'com.example.' + 'o'.repeat(63),
  'com.' + 'middle.'.repeat(40) + 'foo',
  'com.example.fooBar',
  'net.users.bob.ping',
  'a.b.c',
  'm.xn--masekowski-d0b.pl',
  'one.two.three',
  'one.two.three.four-and.FiVe',
  'one.2.three',
  'a-0.b-1.c',
  'a0.b1.cc',
  'cn.8.lex.stuff',
  'test.12345.record',
  'a01.thing.record',
  'a.0.c',
  'xn--fiqs8s.xn--fiqa61au8b7zsevnm8ak20mc4a87e.record.two',
  'a0.b1.c3',
  'com.example.f00',
  'onion.expyuzz4wqqyqhjn.spec.getThing',
  'onion.g2zyxa5ihm7nsggfxnu52rck2vv4rvmdlkiu3zzui5du4xyclen53wid.lex.deleteThing',
  'org.4chan.lex.getThing',
  'cn.8.lex.stuff',
  'onion.2gzyxa5ihm7nsggfxnu52rck2vv4rvmdlkiu3zzui5du4xyclen53wid.lex.deleteThing',
  'a.'.repeat(158) + 'a',
])

bench('invalid NSIDs', false, [
  'a.'.repeat(158) + '9',
  'a.'.repeat(154) + 'aðŸ˜….9',
  'o'.repeat(64) + '.foo.bar',
  'com.' + 'o'.repeat(64) + '.foo',
  'com.example.' + 'o'.repeat(64),
  'com.' + 'middle.'.repeat(50) + 'foo',
  'com.example.foo.*',
  'com.example.foo.blah*',
  'com.example.foo.*blah',
  'com.exaðŸ’©ple.thing',
  'a-0.b-1.c-3',
  'a-0.b-1.c-o',
  '1.0.0.127.record',
  '0two.example.foo',
  'example.com',
  'com.example',
  'a.',
  '.one.two.three',
  'one.two.three ',
  'one.two..three',
  'one .two.three',
  ' one.two.three',
  'com.atproto.feed.p@st',
  'com.atproto.feed.p_st',
  'com.atproto.feed.p*st',
  'com.atproto.feed.po#t',
  'com.atproto.feed.p!ot',
  'com.example-.foo',
  'com.-example.foo',
  'com.example.0foo',
  'com.example.f-o',
])

function bench(name, expectedResult, cases) {
  const validators = {
    parsed: (nsid) => validateNsid(nsid).success,
    regexp: (nsid) => validateNsidRegex(nsid).success,
    optimized: (nsid) => validateNsidOptimized(nsid).success,
  }

  const times = Object.fromEntries(Object.keys(validators).map((k) => [k, 0]))

  for (let i = 0; i < 1000; i++) {
    for (const [name, fn] of Object.entries(validators)) {
      const start = performance.now()
      for (let j = 0; j < 20; j++) {
        for (const value of cases) {
          if (fn(value) !== expectedResult) {
            throw new Error(`Validator ${name} gave wrong result`)
          }
        }
      }
      times[name] += performance.now() - start
    }
  }

  console.log(
    name,
    Object.fromEntries(
      Object.entries(times).map(([k, v]) => [k, `${v.toFixed(2)} ms`]),
    ),
  )
}

/** @param value {string} */
function validateNsidOptimized(value) {
  const { length } = value
  if (length > 253 + 1 + 63) {
    return { success: false, message: 'NSID is too long (317 chars max)' }
  }

  let partCount = 1
  let partStart = 0
  let partHasLeadingDigit = false
  let partHasHyphen = false

  let charCode
  for (let i = 0; i < length; i++) {
    charCode = value.charCodeAt(i)

    // Hot path: check frequent chars first
    if (
      (charCode >= 97 && charCode <= 122) /* a-z */ ||
      (charCode >= 65 && charCode <= 90) /* A-Z */
    ) {
      // All good
    } else if (charCode >= 48 && charCode <= 57 /* 0-9 */) {
      if (i === 0) {
        return {
          success: false,
          message: 'NSID first part may not start with a digit',
        }
      }

      // All good

      if (i === partStart) {
        partHasLeadingDigit = true
      }
    } else if (charCode === 45 /* - */) {
      if (i === partStart) {
        return {
          success: false,
          message: 'NSID part can not start with hyphen',
        }
      }
      if (i === length - 1 || value.charCodeAt(i + 1) === 46 /* . */) {
        return { success: false, message: 'NSID part can not end with hyphen' }
      }

      // All good

      partHasHyphen = true
    } else if (charCode === 46 /* . */) {
      // Check prev part size
      if (i === partStart) {
        return { success: false, message: 'NSID parts can not be empty' }
      }
      if (i - partStart > 63) {
        return { success: false, message: 'NSID part too long (max 63 chars)' }
      }

      // All good

      partCount++
      partStart = i + 1
      partHasHyphen = false
      partHasLeadingDigit = false
    } else {
      return {
        success: false,
        message:
          'Disallowed characters in NSID (ASCII letters, digits, dashes, periods only)',
      }
    }
  }

  // Check last part size
  if (length === partStart) {
    return { success: false, message: 'NSID parts can not be empty' }
  }
  if (length - partStart > 63) {
    return { success: false, message: 'NSID part too long (max 63 chars)' }
  }

  // Check last part chars
  if (partHasHyphen || partHasLeadingDigit) {
    return {
      success: false,
      message:
        'NSID name part must be only letters and digits (and no leading digit)',
    }
  }

  // Check part count
  if (partCount < 3) {
    return { success: false, message: 'NSID needs at least three parts' }
  }

  return { success: true, value }
}
