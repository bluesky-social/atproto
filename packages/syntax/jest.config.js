/** @type {import('jest').Config} */
module.exports = {
  displayName: 'Syntax',
  transform: { '^.+\\.(t|j)s$': '@swc/jest' },
  setupFiles: ['<rootDir>/../../jest.setup.ts'],
}
