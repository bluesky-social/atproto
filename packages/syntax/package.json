{
  "name": "@atproto/syntax",
  "version": "0.1.5",
  "license": "MIT",
  "description": "Validation for atproto identifiers and formats: DID, handle, NSID, AT URI, etc",
  "keywords": [
    "atproto",
    "did",
    "nsid",
    "at-uri"
  ],
  "homepage": "https://atproto.com",
  "repository": {
    "type": "git",
    "url": "https://github.com/bluesky-social/atproto",
    "directory": "packages/syntax"
  },
  "main": "src/index.ts",
  "publishConfig": {
    "main": "dist/index.js",
    "types": "dist/index.d.ts"
  },
  "scripts": {
    "test": "jest",
    "build": "node ./build.js",
    "postbuild": "tsc --build tsconfig.build.json",
    "update-main-to-dist": "node ../../update-main-to-dist.js packages/syntax"
  },
  "dependencies": {
    "@atproto/common-web": "workspace:^"
  },
  "browser": {
    "dns/promises": false
  }
}
