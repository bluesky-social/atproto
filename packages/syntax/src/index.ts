export * from './handle'
export * from './did'
export * from './nsid'
export * from './aturi'
export * from './tid'
export * from './recordkey'
export * from './datetime'
