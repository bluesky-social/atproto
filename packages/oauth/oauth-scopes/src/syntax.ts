import { minIdx, toRecord } from './lib/util'

export type NeArray<T> = [T, ...T[]]

/**
 * Non-empty readonly array
 */
export type NeRoArray<T> = readonly [T, ...T[]]

export type ScopeForResource<R extends string> =
  | R
  | `${R}:${string}`
  | `${R}?${string}`

export type ResourceSyntaxJson<R extends string = string> = {
  resource: R
  positional?: string
  params?: Record<string, undefined | string | NeRoArray<string>>
}

/**
 * Allows to quickly check if a scope is for a specific resource.
 */
export function isScopeForResource<R extends string>(
  scope: string,
  resource: R,
): scope is ScopeForResource<R> {
  if (!scope.startsWith(resource)) return false
  if (scope.length === resource.length) return true

  const nextCharCode = scope.charCodeAt(resource.length)
  return nextCharCode === 0x3a /* : */ || nextCharCode === 0x3f /* ? */
}

/**
 * Utility class to parse and interpret the resource scope syntax used in
 * atproto oauth scopes.
 * The syntax is defined as follows:
 * ```nbf
 * scope := resource [':' positional] ['?' params]
 * params := param ['&' param]*
 * param := name '=' value
 * ```
 * Where "positional" can be used as short-hand (i.e. not used in combination
 * with) for a specific parameter.
 */
export class ResourceSyntax<R extends string = string> {
  constructor(
    public readonly resource: R,
    public readonly positional?: string,
    public readonly params?: Readonly<URLSearchParams>,
  ) {}

  is<T extends string>(resource: T): this is ResourceSyntax<T> {
    return this.resource === (resource as string)
  }

  containsParamsOtherThan(allowedParam: readonly string[]): boolean {
    const { params } = this
    if (params) {
      for (const key of params.keys()) {
        if (!allowedParam.includes(key)) return true
      }
    }

    return false
  }

  /**
   * Retrieve the value of a parameter that only allows a single value. If the
   * parameter is not found, it will return `undefined`. If the syntax is
   * incorrect (i.e. the parameter has multiple values), it will return `null`.
   */
  getSingle(name: string, isPositional = false): string | undefined | null {
    const { params } = this
    const values =
      params != null && params.has(name)
        ? (params.getAll(name) as [string, ...string[]])
        : undefined

    if (!values) {
      // No named parameter found, use positional parameter
      if (isPositional) return this.positional

      return undefined
    }

    if (values.length !== 1) {
      // Single value expected
      return null
    }

    if (isPositional && this.positional !== undefined) {
      // Positional parameter cannot be used with named parameters
      return null
    }

    return values[0]
  }

  /**
   * Retrieve the values of a parameter that allows multiple values. If the
   * parameter is not found, it will return `undefined`. If the syntax is
   * incorrect (i.e. there is bot a positional and named parameter), it will
   * return `null`.
   */
  getMulti(name: string, isPositional?: false): NeRoArray<string> | undefined
  getMulti(
    name: string,
    isPositional: boolean, // Only if this arg is true, will this method return null
  ): NeRoArray<string> | null | undefined
  getMulti(
    name: string,
    isPositional = false,
  ): NeRoArray<string> | null | undefined {
    const { params } = this
    const values =
      params != null && params.has(name)
        ? (params.getAll(name) as [string, ...string[]])
        : undefined

    if (!values) {
      // No named parameter found, use positional parameter

      if (isPositional && this.positional !== undefined) {
        return [this.positional]
      }

      return undefined
    }

    if (isPositional && this.positional !== undefined) {
      // @NOTE we *could* return [this.positional, ...values] here but the
      // atproto scope syntax forbids use of both positional and named
      // parameters.
      return null
    }

    return values
  }

  toString(): ScopeForResource<R> {
    return encodeScope(this.resource, this.positional, this.params)
  }

  toJSON(): ResourceSyntaxJson<R> {
    return {
      resource: this.resource,
      positional: this.positional,
      params: this.params?.size ? toRecord(this.params) : undefined,
    }
  }

  static fromString<R extends string>(
    scope: R | `${R}:${string}` | `${R}?${string}`,
  ): ResourceSyntax<R> {
    const paramIdx = scope.indexOf('?')
    const colonIdx = scope.indexOf(':')

    const resourceEnd = minIdx(paramIdx, colonIdx)

    const resource = (
      resourceEnd !== -1 ? scope.slice(0, resourceEnd) : scope
    ) as R

    const positional =
      colonIdx !== -1
        ? // There is a positional parameter, extract it
          paramIdx === -1
          ? decodeURIComponent(scope.slice(colonIdx + 1))
          : colonIdx < paramIdx
            ? decodeURIComponent(scope.slice(colonIdx + 1, paramIdx))
            : undefined
        : undefined

    const params =
      paramIdx !== -1 // There is a query string
        ? paramIdx === scope.length - 1
          ? undefined // The query string is empty
          : new URLSearchParams(scope.slice(paramIdx + 1))
        : undefined

    return new ResourceSyntax(resource, positional, params)
  }
}

/**
 * Format a scope string for a resource with parameters
 * as a positional parameter, if possible (if it has only one value).
 * @param resource - The resource name (e.g. `rpc`, `repo`, etc.)
 * @param inputParams - The list of parameters.
 * @param positionalName - The name of the parameter that should be used as
 * positional parameter.
 */
export function formatScope<R extends string>(
  resource: R,
  inputParams: Iterable<
    [name: string, value: undefined | string | NeRoArray<string>]
  >,
  positionalName?: string,
): ScopeForResource<R> {
  const queryParams = new URLSearchParams()

  let positionalValue: string | undefined = undefined

  for (const [name, value] of inputParams) {
    if (value === undefined) continue

    const setPositional =
      name === positionalName && positionalValue === undefined

    if (typeof value === 'string') {
      if (setPositional) {
        positionalValue = value
      } else {
        queryParams.append(name, value)
      }
    } else {
      // value is "readonly [string, ...string[]]"
      if (value.length === 0) {
        // This should never happen (because "value" is supposed to be a
        // non-empty array). Because some scope default to "*" (allow
        // everything) when a parameter is not specified, we'd rather be safe
        // here.
        throw new Error(
          `Invalid scope: parameter "${name}" cannot be an empty array`,
        )
      } else if (setPositional && value.length === 1) {
        positionalValue = value[0]!
      } else {
        for (const v of value) {
          queryParams.append(name, v)
        }
      }
    }
  }

  // Fool-proof: If the input iterable defines multiple times the same
  // positional parameter (name), and it ended up being used as both positional
  // and query param, move the positional value to the query params.
  if (positionalValue !== undefined && queryParams.has(positionalName!)) {
    queryParams.append(positionalName!, positionalValue)
    positionalValue = undefined
  }

  return encodeScope(resource, positionalValue, queryParams)
}

export function encodeScope<R extends string>(
  resource: R,
  positional?: string,
  params?: Readonly<URLSearchParams>,
): ScopeForResource<R> {
  let scope: string = resource

  if (positional !== undefined) {
    scope += `:${encodeScopeComponent(positional)}`
  }

  if (params?.size) {
    scope += `?${normalizeScopeComponent(params.toString())}`
  }

  return scope as ScopeForResource<R>
}

/**
 * Set of characters that are allowed in scope components without encoding. This
 * is used to normalize scope components.
 */
export const ALLOWED_SCOPE_CHARS = new Set(
  // @NOTE This list must not contain "?" or "&" as it would interfere with
  // query string parsing.
  [':', '/', '+', ',', '@', '%'],
)

export function encodeScopeComponent(value: string): string {
  return normalizeScopeComponent(encodeURIComponent(value))
}

const NORMALIZABLE_CHARS_MAP = new Map(
  Array.from(
    ALLOWED_SCOPE_CHARS,
    (c) => [encodeURIComponent(c), c] as const,
  ).filter(
    ([encoded, c]) =>
      // Make sure that any char added to ALLOWED_SCOPE_CHARS that is a char
      // that indeed needs encoding. Also, the normalizeScopeComponent only
      // supports three-character percent-encoded sequences.
      encoded !== c && encoded.length === 3 && encoded.startsWith('%'),
  ),
)

/**
 * Assumes a properly url-encoded string.
 */
export function normalizeScopeComponent(value: string): string {
  // No need to read the last two characters since percent encoded characters
  // are always three characters long.
  let end = value.length - 2

  for (let i = 0; i < end; i++) {
    // Check if the character is a percent-encoded character
    if (value.charCodeAt(i) === 0x25 /* % */) {
      // Read the next encoded char. Current version only supports
      // three-character percent-encoded sequences.
      const encodedChar = value.slice(i, i + 3)

      // Check if the encoded character is in the normalization map
      const normalizedChar = NORMALIZABLE_CHARS_MAP.get(encodedChar)
      if (normalizedChar) {
        // Replace the encoded character with its normalized version
        value = `${value.slice(0, i)}${normalizedChar}${value.slice(i + encodedChar.length)}`

        // Adjust index to account for the length change
        i += normalizedChar.length - 1

        // Adjust end index since we replaced encoded char with normalized char
        end -= encodedChar.length - normalizedChar.length
      }
    }
  }

  return value
}
