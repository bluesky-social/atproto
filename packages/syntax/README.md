# @atproto/syntax: validation helpers for identifier strings

Validation logic for [atproto](https://atproto.com) identifiers - DIDs, Handles, NSIDs, and AT URIs.

[![NPM](https://img.shields.io/npm/v/@atproto/crypto)](https://www.npmjs.com/package/@atproto/syntax)
[![Github CI Status](https://github.com/bluesky-social/atproto/actions/workflows/repo.yaml/badge.svg)](https://github.com/bluesky-social/atproto/actions/workflows/repo.yaml)

## Usage

### Handles

Syntax specification: <https://atproto.com/specs/handle>

```typescript
import { isValidHandle, ensureValidHandle, isValidDid } from '@atproto/syntax'

isValidHandle('alice.test') // returns true
ensureValidHandle('alice.test') // returns void

isValidHandle('al!ce.test') // returns false
ensureValidHandle('al!ce.test') // throws

ensureValidDid('did:method:val') // returns void
ensureValidDid(':did:method:val') // throws
```

### NameSpaced IDs (NSID)

Syntax specification: <https://atproto.com/specs/nsid>

```typescript
import { NSID } from '@atproto/syntax'

const id1 = NSID.parse('com.example.foo')
id1.authority // => 'example.com'
id1.name // => 'foo'
id1.toString() // => 'com.example.foo'

const id2 = NSID.create('example.com', 'foo')
id2.authority // => 'example.com'
id2.name // => 'foo'
id2.toString() // => 'com.example.foo'

const id3 = NSID.create('example.com', 'someRecord')
id3.authority // => 'example.com'
id3.name // => 'someRecord'
id3.toString() // => 'com.example.someRecord'

NSID.isValid('com.example.foo') // => true
NSID.isValid('com.example.someRecord') // => true
NSID.isValid('example.com/foo') // => false
NSID.isValid('foo') // => false
```

### AT URI

Syntax specification: <https://atproto.com/specs/at-uri-scheme>

```typescript
import { AtUri } from '@atproto/syntax'

const uri = new AtUri('at://bob.com/com.example.post/1234')
uri.protocol // => 'at:'
uri.origin // => 'at://bob.com'
uri.hostname // => 'bob.com'
uri.collection // => 'com.example.post'
uri.rkey // => '1234'
```

## License

This project is dual-licensed under MIT and Apache 2.0 terms:

- MIT license ([LICENSE-MIT.txt](https://github.com/bluesky-social/atproto/blob/main/LICENSE-MIT.txt) or http://opensource.org/licenses/MIT)
- Apache License, Version 2.0, ([LICENSE-APACHE.txt](https://github.com/bluesky-social/atproto/blob/main/LICENSE-APACHE.txt) or http://www.apache.org/licenses/LICENSE-2.0)

Downstream projects and end users may chose either license individually, or both together, at their discretion. The motivation for this dual-licensing is the additional software patent assurance provided by Apache 2.0.
