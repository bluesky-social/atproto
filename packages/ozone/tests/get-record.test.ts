import {
  SeedClient,
  TestNetwork,
  basicSeed,
  TestOzone,
  ModeratorClient,
} from '@atproto/dev-env'
import AtpAgent from '@atproto/api'
import { AtUri } from '@atproto/syntax'
import {
  REASONOTHER,
  REASONSPAM,
} from '../src/lexicon/types/com/atproto/moderation/defs'
import { forSnapshot } from './_util'

describe('admin get record view', () => {
  let network: TestNetwork
  let ozone: TestOzone
  let agent: AtpAgent
  let sc: SeedClient
  let modClient: ModeratorClient

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'ozone_admin_get_record',
    })
    ozone = network.ozone
    agent = ozone.getClient()
    sc = network.getSeedClient()
    modClient = ozone.getModClient()
    await basicSeed(sc)
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  beforeAll(async () => {
    await sc.createReport({
      reportedBy: sc.dids.bob,
      reasonType: REASONSPAM,
      subject: {
        $type: 'com.atproto.repo.strongRef',
        uri: sc.posts[sc.dids.alice][0].ref.uriStr,
        cid: sc.posts[sc.dids.alice][0].ref.cidStr,
      },
    })
    await sc.createReport({
      reportedBy: sc.dids.carol,
      reasonType: REASONOTHER,
      reason: 'defamation',
      subject: {
        $type: 'com.atproto.repo.strongRef',
        uri: sc.posts[sc.dids.alice][0].ref.uriStr,
        cid: sc.posts[sc.dids.alice][0].ref.cidStr,
      },
    })
    await modClient.emitModerationEvent({
      event: { $type: 'com.atproto.admin.defs#modEventTakedown' },
      subject: {
        $type: 'com.atproto.repo.strongRef',
        uri: sc.posts[sc.dids.alice][0].ref.uriStr,
        cid: sc.posts[sc.dids.alice][0].ref.cidStr,
      },
    })
  })

  it('gets a record by uri, even when taken down.', async () => {
    const result = await agent.api.com.atproto.admin.getRecord(
      { uri: sc.posts[sc.dids.alice][0].ref.uriStr },
      { headers: await ozone.modHeaders() },
    )
    expect(forSnapshot(result.data)).toMatchSnapshot()
  })

  it('gets a record by uri and cid.', async () => {
    const result = await agent.api.com.atproto.admin.getRecord(
      {
        uri: sc.posts[sc.dids.alice][0].ref.uriStr,
        cid: sc.posts[sc.dids.alice][0].ref.cidStr,
      },
      { headers: await ozone.modHeaders() },
    )
    expect(forSnapshot(result.data)).toMatchSnapshot()
  })

  it('fails when record does not exist.', async () => {
    const promise = agent.api.com.atproto.admin.getRecord(
      {
        uri: AtUri.make(
          sc.dids.alice,
          'app.bsky.feed.post',
          'badrkey',
        ).toString(),
      },
      { headers: await ozone.modHeaders() },
    )
    await expect(promise).rejects.toThrow('Record not found')
  })

  it('fails when record cid does not exist.', async () => {
    const promise = agent.api.com.atproto.admin.getRecord(
      {
        uri: sc.posts[sc.dids.alice][0].ref.uriStr,
        cid: sc.posts[sc.dids.alice][1].ref.cidStr, // Mismatching cid
      },
      { headers: await ozone.modHeaders() },
    )
    await expect(promise).rejects.toThrow('Record not found')
  })
})
