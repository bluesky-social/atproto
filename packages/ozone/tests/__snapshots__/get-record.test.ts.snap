// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin get record view gets a record by uri and cid. 1`] = `
Object {
  "blobCids": Array [],
  "blobs": Array [],
  "cid": "cids(0)",
  "indexedAt": "1970-01-01T00:00:00.000Z",
  "labels": Array [
    Object {
      "cid": "cids(0)",
      "cts": "1970-01-01T00:00:00.000Z",
      "neg": false,
      "src": "user(0)",
      "uri": "record(0)",
      "val": "self-label",
    },
  ],
  "moderation": Object {
    "subjectStatus": Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "id": 1,
      "lastReportedAt": "1970-01-01T00:00:00.000Z",
      "lastReviewedAt": "1970-01-01T00:00:00.000Z",
      "lastReviewedBy": "did:example:admin",
      "reviewState": "com.atproto.admin.defs#reviewClosed",
      "subject": Object {
        "$type": "com.atproto.repo.strongRef",
        "cid": "cids(0)",
        "uri": "record(0)",
      },
      "subjectBlobCids": Array [],
      "subjectRepoHandle": "alice.test",
      "takendown": true,
      "updatedAt": "1970-01-01T00:00:00.000Z",
    },
  },
  "repo": Object {
    "did": "user(0)",
    "email": "alice@test.com",
    "handle": "alice.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "invitesDisabled": false,
    "moderation": Object {},
    "relatedRecords": Array [
      Object {
        "$type": "app.bsky.actor.profile",
        "avatar": Object {
          "$type": "blob",
          "mimeType": "image/jpeg",
          "ref": Object {
            "$link": "cids(1)",
          },
          "size": 3976,
        },
        "description": "its me!",
        "displayName": "ali",
        "labels": Object {
          "$type": "com.atproto.label.defs#selfLabels",
          "values": Array [
            Object {
              "val": "self-label-a",
            },
            Object {
              "val": "self-label-b",
            },
          ],
        },
      },
    ],
  },
  "uri": "record(0)",
  "value": Object {
    "$type": "app.bsky.feed.post",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "labels": Object {
      "$type": "com.atproto.label.defs#selfLabels",
      "values": Array [
        Object {
          "val": "self-label",
        },
      ],
    },
    "text": "hey there",
  },
}
`;

exports[`admin get record view gets a record by uri, even when taken down. 1`] = `
Object {
  "blobCids": Array [],
  "blobs": Array [],
  "cid": "cids(0)",
  "indexedAt": "1970-01-01T00:00:00.000Z",
  "labels": Array [
    Object {
      "cid": "cids(0)",
      "cts": "1970-01-01T00:00:00.000Z",
      "neg": false,
      "src": "user(0)",
      "uri": "record(0)",
      "val": "self-label",
    },
  ],
  "moderation": Object {
    "subjectStatus": Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "id": 1,
      "lastReportedAt": "1970-01-01T00:00:00.000Z",
      "lastReviewedAt": "1970-01-01T00:00:00.000Z",
      "lastReviewedBy": "did:example:admin",
      "reviewState": "com.atproto.admin.defs#reviewClosed",
      "subject": Object {
        "$type": "com.atproto.repo.strongRef",
        "cid": "cids(0)",
        "uri": "record(0)",
      },
      "subjectBlobCids": Array [],
      "subjectRepoHandle": "alice.test",
      "takendown": true,
      "updatedAt": "1970-01-01T00:00:00.000Z",
    },
  },
  "repo": Object {
    "did": "user(0)",
    "email": "alice@test.com",
    "handle": "alice.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "invitesDisabled": false,
    "moderation": Object {},
    "relatedRecords": Array [
      Object {
        "$type": "app.bsky.actor.profile",
        "avatar": Object {
          "$type": "blob",
          "mimeType": "image/jpeg",
          "ref": Object {
            "$link": "cids(1)",
          },
          "size": 3976,
        },
        "description": "its me!",
        "displayName": "ali",
        "labels": Object {
          "$type": "com.atproto.label.defs#selfLabels",
          "values": Array [
            Object {
              "val": "self-label-a",
            },
            Object {
              "val": "self-label-b",
            },
          ],
        },
      },
    ],
  },
  "uri": "record(0)",
  "value": Object {
    "$type": "app.bsky.feed.post",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "labels": Object {
      "$type": "com.atproto.label.defs#selfLabels",
      "values": Array [
        Object {
          "val": "self-label",
        },
      ],
    },
    "text": "hey there",
  },
}
`;
